File: cli/azd/extensions/microsoft.azd.foundry/internal/cmd/listen.go
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License.

package cmd

import (
	"context"
	"fmt"
	"time"

	"github.com/azure/azure-dev/cli/azd/pkg/azdext"
	"github.com/spf13/cobra"
)

func newListenCommand() *cobra.Command {
	return &cobra.Command{
		Use:   "listen",
		Short: "Starts the extension and listens for events.",
		RunE: func(cmd *cobra.Command, args []string) error {
			// Create a new context that includes the AZD access token.
			ctx := azdext.WithAccessToken(cmd.Context())

			// Create a new AZD client.
			azdClient, err := azdext.NewAzdClient()
			if err != nil {
				return fmt.Errorf("failed to create azd client: %w", err)
			}
			defer azdClient.Close()

			eventManager := azdext.NewEventManager(azdClient)
			receiveChan := eventManager.Receive(ctx)

			// Register the event handler(s) (synchronously).
			err = eventManager.AddProjectEventHandler(
				"preprovision",
				func(ctx context.Context, args *azdext.ProjectEventArgs) error {
					for i := 1; i <= 20; i++ {
						fmt.Printf("%d. Doing important work in extension...\n", i)
						time.Sleep(250 * time.Millisecond)
					}

					return nil
				},
			)
			if err != nil {
				return fmt.Errorf("failed to add preprovision project event handler: %w", err)
			}

			err = eventManager.AddServiceEventHandler(
				"prepackage",
				func(ctx context.Context, args *azdext.ServiceEventArgs) error {
					for i := 1; i <= 20; i++ {
						fmt.Printf("%d. Doing important work in extension...\n", i)
						time.Sleep(250 * time.Millisecond)
					}

					return nil
				},
				nil,
			)

			if err != nil {
				return fmt.Errorf("failed to add predeploy event handler: %w", err)
			}

			// Block until the Receive function returns, meaning the stream has ended.
			err = <-receiveChan

			return err
		},
	}
}

File: cli/azd/extensions/microsoft.azd.extensions/internal/resources/languages/javascript/commands/listen.js
const { Command } = require('commander');
const AzdClient = require('../azdClient');
const { EventManager } = require('../eventManager');

function createListenCommand() {
  const cmd = new Command('listen');
  cmd.description('Starts the extension and listens for events.');

  cmd.action(async () => {
    const client = new AzdClient();
    const eventManager = new EventManager(client);

    const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));

    await eventManager.addProjectEventHandler('preprovision', async () => {
      for (let i = 1; i <= 10; i++) {
        console.log(`[preprovision] Doing important work... step ${i}`);
        await sleep(200);
      }
    });

    await eventManager.addServiceEventHandler('prepackage', async () => {
      for (let i = 1; i <= 10; i++) {
        console.log(`[prepackage] Doing important work... step ${i}`);
        await sleep(200);
      }
    });

    try {
      await eventManager.receive();
    } catch (err) {
      console.error('Error while receiving events:', err.message);
    }
  });

  return cmd;
}

module.exports = { createListenCommand };
